### ISSUE TITLE
toward P3

### ISSUE THREAD

**Issue by gordonwoodhull** [gereleth/hexapipes#83]
Hi @gereleth!

Since a couple months after you released hexapipes, I've been wondering what it would be like to play this game on [Penrose tiles](https://en.m.wikipedia.org/wiki/Penrose_tiling).

In particular, it looks like P3 with two different sizes of rhombuses would be a good match for this game.

![image](https://user-images.githubusercontent.com/1366709/229299319-1bc6dfb7-16f5-4ffc-bdbf-202f403c28b9.png)

Weird things about P3:
* Every tile has four neighbors, but it's not a grid.
* It is rotationally 5-fold symmetric around exactly one point, and is aperiodic - it doesn't wrap.
* The number of tiles that meet at a vertex is between 3 - 7.

I am hoping that gameplay will also be weird but not broken. It should work, just different maze walls right?

I can only work on this on nights and weekends, so it may take me a while to build this. So I will live-blog my progress in this ticket, rebase regularly, and publish the results from [my fork](https://github.com/gordonwoodhull/hexapipes) at [penrosepipes.vercel.app](https://penrosepipes.vercel.app).

I will of course contribute back once something works, assuming it's still maintainable and makes sense to you. I think it will still be comprehensible. There may be something like a graph abstraction above the grid, the game using the graph instead of the grid.

I've published the first change, which is subtle but important. To handle P3 and other tiles that are not rotationally symmetric, we can imagine arms of different lengths and different angles going to points on the edge of the tile which are on a straight line to the center of the neighboring tile. [^1]

Now instead of rotating the `<g>` using `transform:rotate(...)`, we bitwise rotate the tile, obtain its directions, and return the corresponding deltas.

We can restore animation by applying the CSS transition to the path instead of the transform. 

```css
	.pipe path {
		transition: d 100ms;
	}
```

If you look closely, the animation is not as good, and we might consider adding keyframes so that the arms don't get as short during the animation. But I think it maintains the illusion well enough.

The transition only works on Chrome and Firefox; on WebKit Safari/iOS it degrades to no transition. I am sure there are workarounds for WebKit but I haven't looked into it yet.

`grid.getDirections(...)` already has an optional `rotate` parameter, but there is a wrinkle: we need the arms to always be listed in the same order in order for the animation to make sense. So we also rotate `grid.DIRECTIONS` by the same amount before filtering.

At the current commit (ac09e1c72ca) I haven't yet implemented arm identity for the square grid, and you can see the bug that arises if arm directions are not returned in "original" order. I'll publish the fix for this next week!

[^1]: I hope to find time to draw a diagram showing what arms will look like on P3. It will be weird, because the arms will rotate (and stretch) but the tile will not, but I think it will work.

**Comment by gereleth** [comment]
Hi, Gordon! I'm really excited to see this.


>  It should work, just different maze walls right?

Absolutely! There's [Penrose slitherlink](https://krazydad.com/play/vslitherlink/), why not pipes)).

---

Regarding the rotations of rhombus tiles I thought of implementing them with some skew transform of a square tile. I made a [REPL](https://svelte.dev/repl/d5f992c3fe1949a4b0ef0b5476833b1d?version=3.58.0) as a proof of concept that rotation transitions can look nice in this case. Of course the real difficulty is in figuring out correct transforms for different rhombus types and orientations. But maybe this is still less work than transitioning paths.

I also must confess that I changed a lot about grid implementations while working on "octagons + squares" grid. Basically I abstracted away a `RegularPolygonTile` that handles computing rotations, drawing tiles and pipes, detecting edgemark gestures - all the stuff that's local to the tile. So the grid delegates this local stuff to its polygon tile(s) and stays responsible for "global" questions like:
- where is tile number `i` located (index_to_xy)
- what's the number of tile located at `x, y` point (which_tile_at)
- what's the neighbour of tile `i` in direction `d` (find_neighbour)

Maybe there could be a `SkewedPolygonTile` that "unskews" the coordinates it receives and then reuses regular polygon logic for all the local matters.

I've long wanted to have this "cubes" tiling - implementing this might be a way to try out rhombic tiles without the additional Penrose complications. Grid logic here can reuse the `HexaGrid` and further subdivide each hexagon into three parts.

![Rhombille tiling](https://user-images.githubusercontent.com/6080076/229311804-c7c6cf6d-6066-4157-a141-9f0ce3aa89aa.png)

**Comment by gordonwoodhull** [comment]
I had the related thought of doing some kind of periodic rhombus tiles next week to test the variable length arms and different angles. 

To make it easier I might do the simpler one-orientation rhombus first, even though it’s just squashed and rotated squares.

I had not considered a skew transform—will check out your REPL, thanks! That might be cleaner than the path transitions.

I have not looked at the octagon branch yet, but I will try to rebase on it before proceeding. It’s likely you will merge that before I finish this, and it sounds like you may have implemented some of the same abstractions I’ve been thinking about.

**Comment by gordonwoodhull** [comment]
The disadvantage of skewing a square tile in order to make a rhombus tile is that the pipes will bend at tile boundaries, if the tiles are oriented in different directions. The advantage is that the pipes still land in the center of the tile boundaries.

Hard to say which looks better / clearer without trying.

The new abstractions look very helpful!

**Comment by gordonwoodhull** [comment]
I am taking your advice and starting with cubes. This is... not working yet... but I think it's really close.

https://user-images.githubusercontent.com/1366709/230744040-06016e5b-d989-4d35-900c-f3cb495c8701.mov

**Comment by gereleth** [comment]
I tried to solve one but my clicks land too unpredictably 😂. Individual tiles are looking great though 👍.

**Comment by gordonwoodhull** [comment]
Display and gameplay mostly work for the cubes grid ([demo](https://penrosepipes.vercel.app/cube/5)). 

There are many bugs, among them:
* The generated puzzles are impossible. I haven't looked at the generation code very much, so I'm not sure if it needs to change or if the grid is improperly specified. 
* Clicking on the border between tiles can cause the wrong tile to rotate.
* There is an off-by-one bug in `find_neighbor` where the rows wrap.
* Smudges at tile boundaries are more visible because of the skew. I think the easiest thing to do is to clip each tile, although that may clutter the SVG with `<defs>`.
* I haven't yet tried to get marks to work / looked into `SkewedPolygonTile`. It seemed simpler to start by defining the extra transforms in the grid without changing the polygons.
* The naive non-wrapped grid might be too easy because the border is so bumpy.

Hope to continue this tomorrow.

**Comment by gereleth** [comment]
It's really coming together!

Generation is currently messed up because opposite directions in the grid are inconsistent. The directions are defined like in this image (tile index in black, direction numbers in blue). See how the opposite of direction 2 is sometimes 4 and sometimes 8.
![current directions numbering](https://user-images.githubusercontent.com/6080076/232041967-2e12bc19-3ac2-4d7d-bb76-89770ff93490.png)

You could rotate your squares like this instead. Here directions 1 and 2 look inside the hexagon, and 4 and 8 look outside. This way the opposites are consistent and the generator should make something solvable =).
![possible consistent directions numbering](https://user-images.githubusercontent.com/6080076/232043256-509b2bcd-2f4a-4b00-9a4e-aa06b5fec574.png)

Or you could use the hexagonal directions... But then each of the three squares would have its own set of directions so I don't know if this idea is any good)).

Either way I'm also thinking that maybe we shouldn't rely on `grid.OPPOSITE` constant and instead return opposite direction as part of `find_neighbour` output. This seems like a good thing to hide away inside the grid but will require some slight changes to solver, generator and game logic.

----

For wrong tiles rotating and wrap errors I thought that maybe literally doing `this.hexagrid = new HexaGrid(...)` could be helpful. Then for example you can use `this.hexagrid.which_tile_at(x, y)` to find the hexagon index and only need to look at the angle to tell which rhombus it is. Same with finding neighbours - get neighbour hexagon index first and then figure out where we are inside it. I really don't wish the pain of debugging hexagonal wrapping on anyone =).

----

Boundary smudges may be fixed by setting `stroke-linejoin="bevel"`on pipe insides in `Tile.svelte`. I originally set it to round to avoid these small artifacts seen between connected pipes. But maybe this can be solved some other way.
![image](https://user-images.githubusercontent.com/6080076/232053999-12913861-8ab6-4587-a50f-7dd4f8128b5b.png)

--- 

Re bumpy borders we could always crank up the `avoidObvious` setting).

**Comment by gordonwoodhull** [comment]
Excellent hints and feedback, thanks! 

I agree it would be much better to reuse `HexaGrid` rather than duplicating all this code. I have been expedient to get things working, but I am glad to revise to make this maintainable.

**Comment by gordonwoodhull** [comment]
With the rotation and direction numbering & opposites all consistent, it is now working perfectly in non-wrap mode. (It's playable but difficult to complete in wrap mode, because of the off-by-one bug.)

<img width="745" alt="image" src="https://user-images.githubusercontent.com/1366709/232144467-de80e472-b1fd-405e-a83e-7c904dba956d.png">

I think it's harder than squares but easier than hexagons. The false perspective is disorienting! 😱 

Quite a lot still to fix and clean up - realistically probably won't do a PR until next week.

**Comment by gereleth** [comment]
Wow, I just love solving buggy puzzles. Edge issues on wraps make things more challenging)).

![image](https://user-images.githubusercontent.com/6080076/232189609-60b96349-4a97-40f0-b732-6af2aa8d3b65.png)

Could you add a cubes option to the custom puzzle page? For experimenting with generator settings. I think that this puzzle would definitely benefit from bigger `avoidObvious` values. Maybe you can use the same generation settings as the square puzzles for now? [Defined here](https://github.com/gordonwoodhull/hexapipes/blob/80156d8b641e1c8c287b08c3f24a29fded0b2036/src/lib/puzzleWrapper/PuzzleWrapper.svelte#L175).

**Comment by gordonwoodhull** [comment]
It looks like P3 tiles have the same notion of "opposite sides". From [the wikipedia page](https://en.m.wikipedia.org/wiki/Penrose_tiling#Penrose_tilings), with binary annotation:
![Penrose_rhombs_matching_rules_with_opposite_directions](https://user-images.githubusercontent.com/1366709/232239234-ba6ff9d5-a2cd-4a13-b71e-346fad459113.png)

**Comment by gordonwoodhull** [comment]
`stroke-linejoin: bevel` looks terrible on iOS but it's fine everywhere else, including Safari.

![iOS 5x5 Cube Pipes Puzzle](https://user-images.githubusercontent.com/1366709/232243744-43309b5d-366b-49a8-812e-f515b0f06a77.png)

**Comment by gereleth** [comment]
No worries, have a good trip!

**Comment by gordonwoodhull** [comment]
Cubes are [now completely working](https://penrosepipes.vercel.app/cube-wrap/5), including wrap and edge marks, although connecting marks do not work reliably on touchscreens (walls do). I also need to apply the transform for Orient / Lock mode, which I just tried for the first time.

I got wrap working last week. I was on vacation, so I didn't mention it here. Reusing `HexaGrid` as you suggested removed 70 lines of duplicated code, and fixed wrap and other issues.

Since then I've played more than 100 of the 5 x 5 cube wrap puzzles (which has 75 tiles). I find it quite fun. I think the corners where 3 tiles come together make it slightly easier than square grid, so I'm usually able to prove and auto-lock almost till the end. Best time under 2 minutes but it often takes 10.

I think the non-wrap should use hexagon bounds so that there aren't exposed corners. 

I'm using [transformation-matrix](https://www.npmjs.com/package/transformation-matrix) to invert the tile transformation and detect edge marks in the original coordinate system of the polygon. Hope the extra dependency is okay.

Needless to say, the transformation matrix approach is very general and will apply nicely to P3 tiles.

So neat how the maze generation code works perfectly with no changes for this grid. The only thing that "puzzles" me is that occasionally it draws a fully connected X - does it run out of other ideas? I think this only happens on the non-wrap.

I'll squash my commits and submit a PR tomorrow.

Happy to refactor further. I wasn't sure about the responsibility of controls vs grid for doing transformation - putting the code in the controls seemed less repetitive than adding the transformation matrix stuff to every grid, but maybe there's a better way.

Also the "grid tile" containing three actual tiles, each transformed its own way, is maybe not a good abstraction, just expedient.

**Comment by gereleth** [comment]
I played some 5x5 wraps, they're nice! Quite refreshing after the confusing evilness of [etrat grid](https://hexapipes-git-etrat-gereleth.vercel.app/etrat/15) that I've been working on).

It's neat how this grid has two distinct types of corners (with 3 and 6 squares meeting). All the others I've encountered so far have a constant number of tiles meeting in a corner (3 for hexagons and octagons, 4 for squares, 5 for etrat). Places where 3 tiles meet allow for some solver shortcuts but I don't yet know how to apply them to only some places in the grid.
And imagine if each rhombus were a 2x2 square grid, then we'd have 3, 4 or 6 tiles meeting in different corners :exploding_head: 

Re connection marks I think it would be cool if they could bend at the tile border like the pipes do. I drew a rough example on the pinkish pipe here. But this might be tricky to implement. Like each of the two tiles should draw its own half using its own transforms? Or maybe let's end the mark at tile border at least. The part that's hanging off doesn't look good to me.
![bending connection mark example](https://user-images.githubusercontent.com/6080076/236620786-61f17333-7c56-4603-b6cf-d77585cdb7d1.png)

> I think the non-wrap should use the hexagon playing field shape

There's a helper for that in hexagrid code, something like `grid.useShape('hexagon')`. It makes the corners empty to get a hexagonal field shape. I use that for dailies sometimes). You could iterate over empty hexagons and empty some rhombs in turn.

> The only thing that "puzzles" me is that occasionally it draws a fully connected X 

Fully connected tiles can happen during pregeneration if there is no other way to complete a puzzle. Imagine the red part here is already visited, and the upper left green corner is not. Both neighbouring red tiles are already T-shaped. And the only way to connect to the green corner is to turn one of them into an X. I don't want to introduce backtracking into pregeneration so I just allow this. A less bumpy field shape (like a hexagon) should make X tiles less likely to appear.

![fully connected tile example](https://user-images.githubusercontent.com/6080076/236623259-2f0fbf66-c738-4577-903b-2994b85fd0a9.png)

I haven't reviewed how the transformations work yet, I'll be able to do that once I finish with etrat branch. That grid is forcing me to learn about web workers because ensuring uniqueness takes too long to just leave it in the UI thread =).

**Comment by gordonwoodhull** [comment]
Etrat is tricky! I look forward to hours of figuring out proof patterns there.

I agree that the connection edge marks sticking out are ugly. Wrapping around the edge would be great. I haven't figured out how to get both EdgeMark tiles to draw while only one should record. 

As an experiment, I added a flag to get_edgemark_line to shorten the mark, but this affected the walls as well, making them short and displaced.

<img width="321" alt="image" src="https://user-images.githubusercontent.com/1366709/236647403-d5cc7277-a889-42fa-9f45-f104ac39d59d.png">
 
I've run out of time this week, so I'll leave it like that for now.

I also reverted changes to the controls and created gridutils.js - so the grid has exclusive responsibility of the transform. This is needed for `whichEdge` for Orient / Lock mode and makes sense IMO. However, Orient / Lock still doesn't work, so I'll have to dig into that next week.

I want to fix everything before submitting a PR, so it's another week if not two.

**Comment by gordonwoodhull** [comment]
etrat-wrap is too difficult for me, and now I'm scared P3 is going to be too. 😭

Oh well, can't stop now...

**Comment by gereleth** [comment]
Penrose tiles can't wrap by design so they can't be too bad. I expect something around square or cube difficulty.

I don't quite understand yet what makes etrat so bad. It must be the triangular tiles? With square tiles if you have a turn and know one connection then at least you can say that the opposite side is a wall. That's often handy. With triangular tiles a turn + one its connection gives... mostly nothing. It's kinda similar to low branching hexagonal wraps that are hard too.

One other thing is that etrat puzzles tend to have hundreds and thousands of solutions at even moderate sizes. So even when a puzzle has a unique solution it likely has an obscene number of "almost solutions" that look very very plausible until one last part just doesn't connect. And lookups till you find a contradiction tend to be very long.

**Comment by gordonwoodhull** [comment]
Interesting that etrat has so many solutions!
 
Yes I get stuck on “almost” solutions, even with the 10x10 non-wrap. There will be one extra piece or two closed components. I have no idea how to fix it so I start over or give up.

I think you’re right that P3 should be like cubes only irregular, with 3-7 tiles meeting at a corner.

**Comment by gereleth** [comment]
I tried out the cubes code today and I think the transforms logic is currently a bit wasteful. On every click we compose a transform matrix, then invert it too... That's running a lot of the same computations again and again because we only have three distinct cube faces and so three distinct transforms. It should be enough to compute them only once!

I think a better way to go about this would be to do

```
TOP_FACE = new TransformedPolygonTile(SQUARE, transform_top)
RIGHT_FACE = new TransformedPolygonTile(SQUARE, transform_right)
LEFT_FACE = new TransformedPolygonTile(SQUARE, transform_left)

polygonAt(index) {
  return [RIGHT_FACE, TOP_FACE, LEFT_FACE][index%3]
}
```

The `TransformedPolygonTile` computes transformation matrix, its inverse, and the transform style string in its constructor. And then it does pretty much the same stuff as regular polygon tiles taking care of its own coordinate conversions where necessary.
Then the cubegrid logic can look pretty much the same as other grids with `this.polygonAt(index).doStuff...`and forget that its polygons are weird =).

I would also remove the translate part of the transform and just give each rhombus its own center coordinates instead. This just seems easier to reason about. Not that it's easy! I tried to implement something along these lines but got bogged down trying to understand transform computations.

Can you think of any problems with this approach? I don't really know how many distinct transforms a P3 grid would have. Is there a sane number of them or infinite variety? =)

**Comment by gordonwoodhull** [comment]
I agree the transform should only be computed once per tile “type”, and I like this `TransformedPolygonTile` idea. (I think you mentioned it before but I didn’t understand until now.)

Also transforms should not be computed or used for the grids that do not use this feature. I think I can revert those changes and only keep a few small changes to the grid interface.

As for removing the translate… Maybe? I have often wondered whether it is a good idea to have three tiles per grid position. It was the easiest way to reuse HexaGrid but it’s pretty messy.

For P3, each tile will need its own transform.  There are only two shapes, so the skew and scale will be the same for many tiles, but I doubt it’s worth it to count all the possible rotations of those two shapes. (I’ll check though.) Each tile’s transform and inverse could be computed once. These are 4x4 matrices so it’s not a lot of memory.

**Comment by gereleth** [comment]
Based on my coloring experiment there are 5 kite types and 5 dart types in P3, so 10 distinct transforms in total.

![P3 tiles colored by type and orientation](https://github.com/gereleth/hexapipes/assets/6080076/6849dfb8-b24f-48f0-a626-8819c7d0d380)

Rhomb offsets should be easy to do, I'd precompute `OFFSETS = [{dx, dy}, ...]` for three rhomb types - deltas from hexagon center to rhomb center. And then in `which_tile_at` and `getVisibleTiles` add an offset to hexagon center based on rhomb index. I think it makes sense to keep the grid responsible for tile position and the polygon responsible for tile shape.

**Comment by gordonwoodhull** [comment]
<img width="664" alt="Screen Shot 2023-05-26 at 7 17 36 AM" src="https://github.com/gereleth/hexapipes/assets/1366709/96a14478-9b20-4f83-b957-00e1c639a443">

Okay, I think cubes is ready for review.

**Comment by gordonwoodhull** [comment]
Next step for me is to implement the P3 tiling itself. I think I’ll use the substitution / drill-down method detailed here:

https://www.chiark.greenend.org.uk/~sgtatham/quasiblog/aperiodic-tilings/

I’m more familiar with D3 for SVG, so I’ll probably implement the geometric algorithm standalone, and come back to the tiles in a couple weeks when I have a generator for coordinate and rotation data.

**Comment by gordonwoodhull** [comment]
I have a [demo of producing an approximate number of P3 tiles which intersect a polygon](https://gordonwoodhull.github.io/penrose-fill-polygon/penrose.html).

That Tatham paper says how to keep track of adjacency, and I've implemented coordinates for the Robinson triangles. (You can see them by [adding `?coord`](https://gordonwoodhull.github.io/penrose-fill-polygon/penrose.html?coord) - not sure if these are correct yet.) Tatham describes a simple recursive algorithm to find neighbors using the coordinates, and that will give us coalescing of the triangles into rhombuses and tile adjacency.

At that point the only challenge to Penrose pipes is the neighbours and opposites problem we talked about before. Your solution of the grid returning the opposite when returning a neighbour sounded good and I may give that a try in a couple of weeks.

**Comment by gordonwoodhull** [comment]
I shouldn't jinx myself by talking as if I know what all the problems are. "Patching" the border will also be tricky. 

Take this example:

![Screenshot_20230604_134942_Firefox~2](https://github.com/gereleth/hexapipes/assets/1366709/fd7395bf-55b4-4b13-8fe5-09ab058196ab)

We will prune any half-tiles when we coalesce triangles into rhombuses, but here that will cause such a jagged upper-left corner once the half-blue is removed that we should either drop the red or add in the other half-blue.

**Comment by gordonwoodhull** [comment]
Implemented Tatham’s neighboring triangles in the demo above. Only a few more steps before I can integrate p3 tiles into hexapipes.

**Comment by gordonwoodhull** [comment]
It now finds the rhombuses and their neighbours, and it repeatedly culls any tiles which only have one neighbour. The ending shape can be quite different from the original polygon.

<img width="724" alt="pentagon-penrose-neighbors" src="https://github.com/gereleth/hexapipes/assets/1366709/b6d7a7da-7e59-4839-a3c9-dc0c7a811a71">

It might be nicer to fill gaps around the border, but I haven't gotten that working, so it's hidden. I think the cull strategy works well enough to try this in hexapipes, maybe next week.

**Comment by gordonwoodhull** [comment]
I had to debug the categorization of the ten rhombuses from coordinates, so I recreated your coloring experiment from the generated tile coordinates, binning the rhombuses by their tile-relative coordinates.

<img width="765" alt="image" src="https://github.com/gereleth/hexapipes/assets/1366709/b6c69434-a1b9-4b38-9f08-be09798bd18a">

These numbers/colors are arbitrary - next to choose a consistent numbering.

**Comment by gereleth** [comment]
What does it mean when some rhombs are white and without numbers? This usually happens when the area highlighted in the triangle on top is very small.

![image](https://github.com/gereleth/hexapipes/assets/6080076/135ecf5b-da1e-4791-857e-b2281baff5dd)

I like how the grid is different every time but I just realized we'll also need some way to save and restore the generated grid in addition to puzzle instance and the player's progress.

**Comment by gordonwoodhull** [comment]
Thanks, that's a bug. I had set the precision to 11 digits but it doesn't work when the polygon radius is really small. It seems to be okay at 10.

There isn't any reason to use such high precision, I just wanted to be sure that rhombs generated with trig are the same as rhombs generated by triangle decomposition. 

Technically the algorithm is deterministic and we could regenerate the same grid from the polygon shape, center and radius. You can see this by clicking the "fix" link. But it's probably safer to save the grid, which consists per tile of location, base rhombus [^1], tatham coordinates and optionally neighbours.

[^1]: which polygon out of 5 rotations and two shapes - I haven't found a good name for this, "face" was nice for cubes but doesn't make sense here. "base" is ok.

**Comment by gordonwoodhull** [comment]
Implementing `P3Grid` was easy with the new `AbstractGrid`. Still much debugging to do, and grid storage, but I think we might see something next week or the week after.

I've been calling it P3 because there are other Penrose tilings and in theory you could play with kite & dart tiles, but what do you think? Is it better just to call it Penrose everywhere?

**Comment by gereleth** [comment]
This is so exciting! I think darts and kites will not work with pipes because the shapes are not symmetric. So calling it penrose should be fine I guess?

I got them to display at least, they look so drunk :joy: 
![5x5 penrose example puzzle](https://github.com/gereleth/hexapipes/assets/6080076/222eb75b-2d42-4d86-857a-0f8ffa6ad9b4)
My clicks do nothing for some reason but I scribble-solved it to verify that the generator did something sensible))
![image](https://github.com/gereleth/hexapipes/assets/6080076/c963309d-5876-4502-939c-f2c354a8376f)


Grid construction is currently non-deterministic, right? There is a problem that the generator component and the page create grids independently and they end up different, not even the number of tiles matches. I temporarily switched to using a generator directly on custom puzzle page to get to this result. Here's a patch with my changes if it helps.

[debug_penrose.patch.txt](https://github.com/gereleth/hexapipes/files/12062797/debug_penrose.patch.txt)

**Comment by gordonwoodhull** [comment]
Wow you got it to work!! Amazing how quickly this fit together - I did some research but basically started the hexapipes integration 24h ago. Less than 200 lines woot.

It looks completely daft with the false perspective. I think we might need to try pipe arms at arbitrary angles instead of skew, but maybe we should keep this for the psychedelic (or as you say “drunk”) effect.

Yes I was “puzzled” why I was seeing different counts & figured it was because I hadn’t dealt with persistence. I didn’t realize there were two grid instances. It’s determinations but it chooses a random polygon to start, we’ll have to decide what state to copy

Thanks for your temporary workaround. So cool!!!

I think I know how I messed up click detection, maybe I can get it running later today.

**Comment by gordonwoodhull** [comment]
I applied your workaround and used the correct coordinates, and it is almost playable, however it has a serious bug which would make it very difficult to complete.

<img width="625" alt="image" src="https://github.com/gereleth/hexapipes/assets/1366709/bd2e185d-49eb-4fd2-847a-55691f10482e">

It's [live here](https://penrosepipes-git-p3-gordonwoodhull.vercel.app/custom).

I have run out of time for this week. If you want to work on it, you are welcome, otherwise I'll continue on Friday.

Although the puzzle is solvable, the display rotation is wrong. Side 1 of a rhombus could be in one of two places, green outline in the penrose-fill-polygon demo:
<img width="150" alt="Screen Shot 2023-07-15 at 8 25 57 PM" src="https://github.com/gereleth/hexapipes/assets/1366709/44b43ca9-ca64-49d7-b3e2-c6ace7c0d8d5">
<img width="131" alt="Screen Shot 2023-07-15 at 8 26 15 PM" src="https://github.com/gereleth/hexapipes/assets/1366709/93d23eca-2892-463c-bfad-3d8a7b1b71c6">

This suggests that there needs to be an offset on the original rotation based on the generated data.

But in playing with it, it looked like some tiles were off by 90º as well. I wouldn't be surprised if angles are reversed somewhere, since I didn't try to account for increasing y up or down, always the bane of trig on the web.

I don't know. There are probably multiple bugs.

I solved a tiny one:

<img width="454" alt="image" src="https://github.com/gereleth/hexapipes/assets/1366709/84174ec4-7c40-45d6-9dff-8ac2d82a381f">

**Comment by gereleth** [comment]
I tried to understand what's going wrong with connections coloring by annotating tiles with `index_base` and clicking around. In the pic below there are two tiles with base 0, they both report the fat red dot as direction 1 and they both say that their neighbour in this direction is tile 6. Obviously one is correct and the other one is turned 180 deg. So the transformed polygon for tile 5 should have directions [4,8,1,2] instead of [1,2,4,8] I guess. I couldn't find what makes these two tiles different. 

![Selection_025](https://github.com/gereleth/hexapipes/assets/6080076/d6e28c78-6c99-4956-8c67-aa2dd55f2f21)

I overlaid my solution to one puzzle (in green) with solver solution (in half-transparent yellow) and we clearly agree on some tiles but disagree by 180 or 90 deg on others. I'm afraid it's up to you to figure out which rotation offsets to use on which tiles...

![Screenshot from 2023-07-16 11-09-33](https://github.com/gereleth/hexapipes/assets/6080076/2ac2fe3f-f2be-4ffa-bb67-63d42a91845f)

**Comment by gordonwoodhull** [comment]
Thanks! Neat diagram and good analysis. 

The orientation comes from the way tiles are generated - as triangles are split this introduces rotation. I couldn't find any pattern to it either, but I thought it made sense to use that side numbering, since neighborhood falls out of it.

I'll figure it out next week. Day job is pretty demanding, so unfortunately I can't completely wear myself out over the weekend.

**Comment by gereleth** [comment]
I couldn't let it go :sweat_smile:  and made a hack for rotation offsets. It works like this:

- draw an edgemark extending slightly out of the tile, use `which_tile_at` to find the neighbour we ended up in
- find that neighbour in the rhomb's neighbours
- compute an offset such that `(edgemark direction index + offset)%4 = neighbour_index`
- remember the offset and use it in subsequent `find_neighbour` calls

This approach is kinda like using a walking stick to check the way forward. I'd much rather see the offsets generated along with the rest of the grid but at least the puzzle is solvable now with consistent connections coloring! See it here: https://hexapipes-git-p3-gereleth.vercel.app/custom

![image](https://github.com/gereleth/hexapipes/assets/6080076/858ee438-9680-454b-a6f3-a8c5964db0f9)

We'll definitely need a better implementation for `which_tile_at` because it's currently linear in the number of tiles and will be too slow in larger puzzles.

Generating a new puzzle hangs the tab pretty often, I haven't investigated why this happens. Edgemarks don't work.

**Comment by gordonwoodhull** [comment]
Wow! It's just as weird as I hoped it would be. Thanks for getting it working! 

Not terribly difficult - I was able to solve a "15x15" with proofs and assistant. I only paused to think once or twice, and didn't have to look for any patterns.

A "haunted" feel to it, because of sudden corners and distorted space. 

Can't wait to fix all this stuff next week. Great fun!

**Comment by gordonwoodhull** [comment]
Working on a "23x23" now.

![Screenshot_20230716_170028_Firefox](https://github.com/gereleth/hexapipes/assets/1366709/1d5c76ed-35a6-40cc-9575-a53ca45c9cbc)

Thoughts:
- I want my edge marks.
- Border is too easy, needs to be cleaned up more. Also I think it fits better in a pentagon than a square.
- It's fun.

**Comment by gordonwoodhull** [comment]
Wow cool!

So neat how the edge marks work just as you would expect, even at those topsy-turvy angles. 

My conclusion was that there *should be* consistent opposite sides but just not in the same order for the two shapes.  

That's what is shown in this diagram, which illustrates how rhombs are allowed to fit together: 

![232239234-ba6ff9d5-a2cd-4a13-b71e-346fad459113](https://github.com/gereleth/hexapipes/assets/1366709/493f0af0-316d-438e-8f4a-e4748d30c67c)

I don't think there's any harm to doing it the way you describe. That's a tiny amount of space and calculation.  

Maybe it will become more clear if we get the original sides 0-3 mapped correctly to 1, 2, 4, 8 without the "walking stick" and offsets/maps. But probably not. 🤪

**Comment by gereleth** [comment]
I tried to paint these jigsaw-like marks on the edges of one puzzle. Here's what I found out:

- at first it seemed like there could be 4 possible assignments (take one edge, give it an arbitrary mark, derive the rest from there). But only one of them actually worked - three others needed tiles different from the two pictured above.
- it looks like there are not 10 but 20 base polygons - original 10 + the same ones turned 180 deg. See for example four light-green base 0 polygons and one dark-green that is rotated compared to them. Same with other light-dark color pairs.

So if we could somehow find this magic directions assignment and use 20 base polygons then we'd have consistent opposites and no need for edgemark hacks. Can we track these jigsaw edge types during grid construction?..

Polygons can use arbitrary directions - like `[1,2,8,4]` or `[8,4,2,1]`. But which to use where is the real question here.

![image](https://github.com/gereleth/hexapipes/assets/6080076/30bf22d5-c163-430d-9a82-54b75f04f4c0)

**Comment by gordonwoodhull** [comment]
Nice. I see what you are saying - yes it is 20 polygons if you consider the 180° turns.

The jigsaw edge types should be inherent to the way the triangles / rhombuses are generated. It’s all based on Tatham’s combinatorial coordinates. (Links in [the readme](https://github.com/gordonwoodhull/penrose-fill-polygon#readme).)

When producing the rhombuses from triangles, side 0 of the two triangles is pasted together. For the narrow rhombus, one triangle will be C and one D. For the thick rhombus, one triangle will be X and one will be Y - and the paste-edge is the long diagonal instead of the short.

Then I use the triangle sides in order X1, X2, Y1, Y2. Or C1, C2, D1, D2.

But I haven’t checked how this corresponds to the jigsaw edges in Wikipedia. I’m pretty sure there is a consistent mapping, but I simply haven’t looked.

The long-diagonal paste vs short-diagonal paste may explain a 90-degree turn. I.e. if the triangle-paste line is horizontal then the skinny rhombus will point up-down instead of left-right.

Hope this helps… gotta run for now. Thinking I may take Friday off to try to knock this out. I’m getting pulled into some demanding projects and I’m not sure if I’ll be able to spend weekends on this going forward. 😔

**Comment by gereleth** [comment]
I looked into why the puzzle page hangs sometimes and the reason is likely that the grid creates detached tiles like this:

![image](https://github.com/gereleth/hexapipes/assets/6080076/a2fdf071-b605-4924-b5b6-1ed0ab5e9e0c)

Disjoint grids have not come up before so the generator has no protections against them)). This looks like something best avoided during grid construction.

**Comment by gordonwoodhull** [comment]
Aha, makes sense. I’m glad there is a simple answer to that. I’ve been playing quite a bit but I can’t get the page back when it hangs like that on my tablet. 

Yes the generator absolutely should not allow disjoint tiles to get through. Wish it didn’t produce any tiles of degree 1 either…

**Comment by gordonwoodhull** [comment]
I have implemented persistence of the grid, and the regular sizes work.

The generator produces the grid and sends it back. Then grid gets put alongside tiles in local storage, both in instance and progress.

This only happens if the grid supports `getState()` and `initialize()` - I implemented this in a non-typesafe way but I'm happy to revisit this.

There is one inefficiency - `P3Grid` still calculates a Penrose tiling when it is constructed. It would be better to call `initialize()` after `createGrid` but there are 8 places `createGrid` is called, so I held off on that.

As for the slowness of `which_tile_at`, I did some measurements, and it only gets into milliseconds once there are hundreds of tiles. With "100x100" which is implemented as "minimum of 10000" and producing 18000-24000 tiles, I was seeing `which_tile_at` take as much as 3-4 milliseconds.

This is fine for interactive on-click use, but it becomes a problem when multiplied by the walking stick poking around every tile. Might look at that next.

**Comment by gereleth** [comment]
Nice, I played some already. Even got some quasi-wrap experience by starting a large puzzle from the middle)).

> Then grid gets put alongside tiles in local storage, both in instance and progress.

I don't get why it needs to be in progress, seems to me it should live in instance only. I haven't looked at the code yet but I'll review how `createGrid` should work with different-every-time grids.

---

Detached tiles strike again and my streak is dead. It was only six games long  :smiling_face_with_tear:. At least the tab didn't hang).

I also saw an example where two thin rhombs were detached from the rest of the grid. So we can't even use "tile with no neighbours" logic to detect this situation. I thought of declaring these tiles empty if the generator can't reach them. But sometimes the generator will choose to start from a detached tile and then the whole grid ends up empty. This is exactly what happened in the previous detached example. Tricky stuff)).

![image](https://github.com/gereleth/hexapipes/assets/6080076/fdc58220-d53a-46c5-9c2d-5ba75c04aac7)

**Comment by gordonwoodhull** [comment]
You're right, the progress state doesn't need grid. 

I've switched to cull strategy, not as nice border in some cases but eliminates islands by repeatedly removing any rhombs with less than two neighbors.

Chose the harder tiles, renamed to [penrose](https://penrosepipes-git-p3-gordonwoodhull.vercel.app/penrose/5).

I think the walking stick is the major problem left. I'm looking at it now.

**Comment by gordonwoodhull** [comment]
<img width="750" alt="image" src="https://github.com/gereleth/hexapipes/assets/1366709/a60021bf-63b6-4a3e-9244-7a3f4a19c349">

**Comment by gordonwoodhull** [comment]
debugging

<img width="819" alt="rotation debug 1" src="https://github.com/gereleth/hexapipes/assets/1366709/aebb1df6-be57-4e0e-8799-1bbc729ec8b5">

<img width="797" alt="rotation debug 2" src="https://github.com/gereleth/hexapipes/assets/1366709/d01a1e5b-2460-48ba-9d62-f244ffe620e1">

<img width="1305" alt="rotation debug 3" src="https://github.com/gereleth/hexapipes/assets/1366709/d60f8dcb-1796-44b3-9d82-e08a111ae39d">

<img width="786" alt="rotation debug 4" src="https://github.com/gereleth/hexapipes/assets/1366709/80c60cc1-2046-42a9-aacb-6bc59d0ae0e4">

**Comment by gordonwoodhull** [comment]
<img width="827" alt="image" src="https://github.com/gereleth/hexapipes/assets/1366709/7fa331ad-0481-43ee-8ab5-ef47ecd9adc8">

**Comment by gordonwoodhull** [comment]
> Even got some quasi-wrap experience by starting a large puzzle from the middle)).

In theory, I think you could have an infinite penrose pipes puzzle if the maze were generated on the fly as area is explored. Would be a huge endeavor though.

**Comment by gordonwoodhull** [comment]
In the new numbering, the flips are +10 on the old numbering (which was previously implemented incorrectly due to y increasing downward).

<img width="125" alt="p3-tile-numbering" src="https://github.com/gereleth/hexapipes/assets/1366709/09267b4f-ae8e-4355-b00a-a61b8f0c790e">

With flip accounted for, the skinny rhombs are offset 0; the thick rhombs are offset 1.

<img width="140" alt="image" src="https://github.com/gereleth/hexapipes/assets/1366709/23709d4d-3d37-440e-8f3d-c91ae002245b">

**Comment by gordonwoodhull** [comment]
Without cleaning anything else up, I'm pushing the simplified offset calculation. It is fast, puzzle now loads in 1.5s for 100x100 = 18k tiles

Have to fix the example. Otherwise I think it's pretty solid.

Probably should fix the offset problem on the penrose-fill-polygon side but I still don't completely understand. 

Cull can do some stupid stuff on the border:

<img width="595" alt="image" src="https://github.com/gereleth/hexapipes/assets/1366709/14e7dd1e-a3bf-4081-b29c-d3cf18f99eb7">

Some combination of fill + cull is probably the answer.

**Comment by gereleth** [comment]
So can we assign directions with consistent opposites now?

**Comment by gordonwoodhull** [comment]
We could. Since the order of sides in this sense is different for thin vs thick rhombs, there would have to be a mapping for one of them.

On the other hand, what you did with the lesser index seems perfectly reasonable for any grid where you don't know the opposites, so maybe that's worth holding onto.

**Comment by gereleth** [comment]
Super glad to be rid of the walking stick! I hate the hack with inconsistent opposites too to be honest and I want to remove it. But I haven't managed to get consistent directions with the current code.

There must be a bug with determining base polygons because when I tried jigsaw marks again I got this result. See two yellow stars made of fat rhombs. Base 0 and 10 rhombs correctly tell that they're flipped versions of each other. But the rest (base 1, 3, 4, 12) just think they are identical when they should be flipped.
![penrose_base_bug](https://github.com/gereleth/hexapipes/assets/6080076/fce4e1bc-211e-44c9-b24c-e5625025daa5)

Instance data for debug purposes: [penrose_5x5.json.txt](https://github.com/gereleth/hexapipes/files/12137204/penrose_5x5.json.txt). Downloading a puzzle should give this data but that's another todo along with importing a puzzle on custom page)).

I noticed that rhombs data has a `key` property that contains a string of eight numbers. If that key is only for use in `getVisibleTiles` then we can just use index instead. Using index doesn't work in wrap puzzles because we may see multiple copies of the same tile. But here the index should be unique enough since penrose grid doesn't wrap.

**Comment by gordonwoodhull** [comment]
I agree, by the jigsaw marks those stars should be consistently 0 - 4 or 10 - 14. Instead the pieces seem to be interchangeable. 

Why then does the simple offset calculation work? 

That `key` data isn't used by `PenroseGrid`; it's an artifact of classifying the rhombs by base.  `PenroseGrid.getVisibleTiles` uses `coord`, from the tatham coordinates of the rhomb's triangles, but you're right it could just as well use index.

I'm out of time for this week. I'm sure there are many bugs but it runs nice.

**Comment by gordonwoodhull** [comment]
I think I have an idea about the wrong rhomb numbers. I assumed that tatham coordinates would correspond to the jigsaw edges but this is apparently not the case. Will look into it next week. 

Also noted about saving the grid / geometry with the tiles to download. 

It's too bad the data is so bulky. For the example I had to put it in a separate file because it's much larger than `grids.js`.

It's useful to have redundancy for various calculations, but I think the saved form could deduplicate.

**Comment by gordonwoodhull** [comment]
It might sound like an overreaction, but I’ve taken down these deployments for now.

On Saturday I played for a couple hours after I was too tired to work (exploring a 40x40). 

I think the perspective distortion made me ill: it gave me vertigo for a few hours after. Like I couldn’t tell which way was up. Not fun. And it comes back if I think about those twists and turns.

Granted I’m susceptible to motion sickness from video games, but I don’t want to put something out that could harm people.

I want to try somehow making the pipes of uniform dimension. I.e. the tiles still skewed and rotated but the pipes not.

So let’s hold off on this a bit.

**Comment by gereleth** [comment]
So sorry this happened to you! I didn't have such severe symptoms myself but I think I can understand the feeling. I do have to blink and shake my head occasionally to avoid dizziness while playing largish penrose pipes. The chaotically crumpled space is not easy on the eyes.

Displaying pipes without perspective distortion is not without problems either... It's gonna be really tricky to get them to connect smoothly at weird angles. And the pipes change shape on rotation - this is especially noticeable on thin rhombs.
(Please disregard the coloring in my screenshot, I only made the geometry look somewhat correct)
![image](https://github.com/gereleth/hexapipes/assets/6080076/a402127f-ed57-407b-bcb8-de8b0bfb5d96)

**Comment by gordonwoodhull** [comment]
Hi @gereleth! I combined the technique from the very beginning of this post with clip polygons and have a demo available of Penrose Pipes with uniform pipe widths:

![Screenshot from 2023-08-26 17-23-45](https://github.com/gereleth/hexapipes/assets/1366709/e6029015-09dc-425a-a31d-4043ee4588ab)

I suspect that I solved this wrong, not that there is a bug in the maze. But I can't figure this one out.

Now the shapes are completely unfamiliar, but I think this is fair - what do you think?

https://penrosepipes-git-p3-gordonwoodhull.vercel.app/penrose/5

It has one severe bug where it is unable to draw arms that go off the grid. If you look closely you will see another glitch implied by arms going straight to neighbour but clipped by polygon. :wink:

**Comment by gordonwoodhull** [comment]
Haha the edge marks need to be direct and not transformed, too

![image](https://github.com/gereleth/hexapipes/assets/1366709/36fa6e0e-b768-4dd2-a987-1d5cb36591e7)

**Comment by gordonwoodhull** [comment]
After playing a bunch of rounds, I'd say it's confusing but not disorienting: I'm not getting the vertigo I got from the previous version. 

But not having familiar or consistent piece shapes makes it more difficult. If the stats are correct it's taking me 5min for a puzzle that took 2min in the previous version. 

Lots of work before I can release this, even more work before it would be ready for PR.

**Comment by gereleth** [comment]
It's confusing all right, I can't tell a straight piece from a turn at a glance!

You're drawing the pipe line to the center of neighbour, right? In my attempt above I was trying to go for edge middle. Maybe these two approaches could be combined - like we go for edge middle first and then towards neighbour center to ensure a smooth connection. That might make straights mostly straight (with some bendy ends).

In your buggy-looking puzzle the tile on the left of the green deadend looks like it still has disconnects. So it's probably a sneaky T looking outside :smile:.

**Comment by gordonwoodhull** [comment]
Definitely worth trying. I don't like adding extra bends but if it helps the pieces be more recognizable, that would be nice. 

I may also experiment with splines as it's a very wavy layout already. 

And yes that must have been a sneaky T. I am often getting tripped up by 3s on the border. I think every other piece is unambiguous there.

**Comment by gordonwoodhull** [comment]
Okay! I think this looks pretty nice.

<img width="669" alt="image" src="https://github.com/gereleth/hexapipes/assets/1366709/1e755f84-c3f2-4110-8388-fc8441da6f61">

https://penrosepipes-git-p3-gordonwoodhull.vercel.app/penrose/7

There is an optical illusion that the narrow rhombs have swollen pipes, but I have measured it and it's not real, just one of those standard optical illusions.

It has a knob `symbol_portion` which sets how far to draw the recognizable tile shape before then heading toward the counterpart on the next tile. The above is 0.7, which I think is a nice compromise.

Zero draws directly to the center of the next tile, as in last week's iteration.

Values larger than 1 draw past the edge without splining (just for reference):

![image](https://github.com/gereleth/hexapipes/assets/1366709/006d5e7f-35d8-4fa7-b24e-2d1fbb93bd47)



Lower numbers get more wavy:
<img width="587" alt="image" src="https://github.com/gereleth/hexapipes/assets/1366709/9a861eac-cf4a-477f-a68f-79940b59c885">

It's not bad without Beziers, but a a little jittery:
<img width="545" alt="image" src="https://github.com/gereleth/hexapipes/assets/1366709/3d977e68-bdae-4d5e-8342-a9dd8b7fdfa0">

https://penrosepipes-git-p3-gordonwoodhull.vercel.app/penrose/7

I'm halfway to fixing the edge pieces; many other things to fix but I think it's quite playable now.

**Comment by gereleth** [comment]
The 0.7 variant looks awesome, I love it! The shapes are recognizable and the connections are smooth. Being closer to edge middles got rid of corner glitches too. I haven't looked at the code but the results are very nice =).

**Comment by gordonwoodhull** [comment]
Thanks! 

I still need to clean up,  putting the changes safely under flags so they don't break the other grids. In particular clip paths should be enabled only for this grid and for iOS to fix #91.

And there is a backlog of things to fix in the Penrose tiles generator that we have discussed before. (E.g. I had to reverse the list of rhombus points and then offset differently for thick vs thin rhombuses, the equivalent of your `neighbor_idx` workaround.)

Given this will take some time, but I'm eager to get this out in the world, here's my plan:

1. Fix all gameplay bugs.
2. Launch at penrosepipes.vercel.app and announce on HN, X, etc, making it clear that this is a prototype and my intention is to clean it up and merge to hexapipes.
3. Deal with less urgent but annoying design problems like the inconsistencies in the generator, and any issues you raise in code review. Respond to feedback from players. Make the generator a proper package and publish it to npm.
4. File a PR.
5. Assuming we agree on something that can be merged, penrosepipes.vercel.app can serve in the long term as a staging area for further improvements. I'd like to improve the ragged edge resolution, think the splines could be even smoother, etc.

Given my time constraints, this will take months. 

I'm so happy with how it turned out. The aesthetic is something greater than what I and you put into it - reminds me of fanciful fonts from the 19th century, with just a touch of psychedelia.

**Comment by gereleth** [comment]
I made a quick fix for the missing legs problem in the `p3` branch. The logic is to just keep the same direction if we don't have a neighbour to bend for =). My commit touched a lot of lines but the only meaningful changes were around line 220 (the rest was auto formatting on save). I think this looks ok ([preview link](https://hexapipes-git-p3-gereleth.vercel.app/penrose/7)).

A more sophisticated approach could use invisible neighbours for bend info but that would be overkill imho =). 

![image](https://github.com/gereleth/hexapipes/assets/6080076/1263f7e5-fe08-40b0-aeb9-7f2c7ad90cb9)

**Comment by gordonwoodhull** [comment]
That's a nice solution! Indeed I was trying to look for the invisible neighbors, which is not necessary here. In fact I think it's better to go straight to emphasize that there is nothing there. 

I haven't tried slowing down the animation yet, but I think it will transition fine, since there are the same number of points per arm (and it's not trying to change between line and curve). 

I've been playing a lot this week and it really picks up in difficulty on larger grids. This means that if you are playing at your limit, the variation in grid sizes (from a little larger than NxN tiles to almost 3xNxN) can make the difficulty and play times of a particular size erratic. 

So I think I will shrink the polygon after generation and throw out tiles until the size is closer to NxN. Seems wasteful but it should be fast since I can use binary search and the triangle intersection code that I got from SO is wicked fast.

**Comment by gordonwoodhull** [comment]
I'm tired this weekend, so I've been messing with ragged border resolution rather than fixing the other grids.

There was an obvious (in hindsight) bug in the `fill` algorithm, where it was checking for a string suffix incorrectly:

```diff
-            tri => find_tris.some(find => find.indexOf(tri.coord) === find.length - tri.coord.length),
+            tri => find_tris.some(find => find.endsWith(tri.coord)),
```

With that corrected, the fill algorithm delivers less portion of tiles with two sides on the border 80% of the time, across many grid sizes. The drawback is that it isn't specifically looking at degree, so it delivers grids with some tiles with three sides on the border 2-6% of the time (worse for small grids). So we'll still need to run cull after fill.

Typical cull:
<img width="754" alt="Screen Shot 2023-09-16 at 5 54 22 PM" src="https://github.com/gereleth/hexapipes/assets/1366709/461b6eba-4059-4529-b3a0-f9ef5e63634f">

versus fill:
<img width="705" alt="Screen Shot 2023-09-16 at 7 03 28 PM" src="https://github.com/gereleth/hexapipes/assets/1366709/bd9b15aa-8813-4811-a955-ebb934506293">

I've been playing a lot of Penrose Pipes at many sizes, and I remain very happy with the gameplay and the aesthetic.

It will be fun to see the more difficult grids your solver produces. The dailies are very hard even with the cube grid, whereas Penrose is pretty easy with the current settings. Have to go to really big grids to get stuck at all in "proofs mode".

**Comment by gordonwoodhull** [comment]
Hi @gereleth - do you have linting or formatting tools I should apply so that merges are less messy?

**Comment by gereleth** [comment]
I use prettier for formatting, `npm run format` should do it. I normally just have VSCode format files on save.

**Comment by gordonwoodhull** [comment]
Okay, soft launch of [penrosepipes.vercel.app](penrosepipes.vercel.app) is live!

Please test to see if you find any more issues. I'll announce later this week.

**Comment by gordonwoodhull** [comment]
Posted on HN and X, please upvote: https://news.ycombinator.com/item?id=37842719

I'll probably repost on Sunday morning.
